File: Lernwerk/App/AppSettings.swift
import Foundation

final class AppSettings: ObservableObject {
    private enum Keys {
        static let tutor = "llm.tutor"
        static let plan = "llm.plan"
        static let demoMode = "demoMode"
    }

    /// Model for the help panel and the flashcards created from it.
    @Published var tutor: ModelSelection {
        didSet { store(tutor, forKey: Keys.tutor) }
    }

    /// Model that reads the material and builds the study plan.
    @Published var plan: ModelSelection {
        didSet { store(plan, forKey: Keys.plan) }
    }

    @Published var demoMode: Bool {
        didSet { defaults.set(demoMode, forKey: Keys.demoMode) }
    }

    @Published private(set) var providersWithKey: Set<LLMProvider>

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        tutor = AppSettings.load(Keys.tutor, from: defaults) ?? .defaultSelection(for: .tutor, provider: .nvidia)
        plan = AppSettings.load(Keys.plan, from: defaults) ?? .defaultSelection(for: .plan, provider: .openRouter)
        demoMode = defaults.bool(forKey: Keys.demoMode)
        providersWithKey = Set(LLMProvider.allCases.filter { KeychainStore.load(account: $0.keychainAccount) != nil })
    }

    var hasAnyKey: Bool {
        !providersWithKey.isEmpty
    }

    func hasKey(for provider: LLMProvider) -> Bool {
        providersWithKey.contains(provider)
    }

    func saveKey(_ key: String, for provider: LLMProvider) -> Bool {
        let trimmed = key.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, KeychainStore.save(trimmed, account: provider.keychainAccount) else {
            return false
        }
        providersWithKey.insert(provider)
        // Saving a key means the student wants real answers; a demo mode left on from the sample material would hide them.
        demoMode = false
        return true
    }

    func deleteKey(for provider: LLMProvider) {
        KeychainStore.delete(account: provider.keychainAccount)
        providersWithKey.remove(provider)
    }

    func selection(for task: LLMTask) -> ModelSelection {
        switch task {
        case .tutor: return tutor
        case .plan: return plan
        }
    }

    func modelLabel(for task: LLMTask) -> String {
        if demoMode { return "Demo" }
        let chosen = self.selection(for: task)
        return chosen.provider.option(for: chosen.model)?.name ?? chosen.model
    }

    func makeClient(for task: LLMTask) -> any LLMClient {
        if demoMode {
            return DemoLLMClient()
        }
        let chosen = self.selection(for: task)
        let model = chosen.model.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !model.isEmpty else {
            return FailingClient(error: .missingModel)
        }
        guard let key = KeychainStore.load(account: chosen.provider.keychainAccount) else {
            return FailingClient(error: .missingAPIKey(provider: chosen.provider.name))
        }
        switch chosen.provider {
        case .anthropic:
            return ClaudeClient(apiKey: key, model: model)
        case .nvidia, .openRouter:
            let fallback = chosen.provider.fallbackModelIDs.first { $0 != model }
            return OpenAICompatibleClient(
                provider: chosen.provider,
                apiKey: key,
                model: model,
                sendsImages: chosen.sendsImages,
                fallbackModels: fallback.map { [$0] } ?? []
            )
        }
    }

    private func store(_ selection: ModelSelection, forKey key: String) {
        if let data = try? JSONEncoder().encode(selection) {
            defaults.set(data, forKey: key)
        }
    }

    private static func load(_ key: String, from defaults: UserDefaults) -> ModelSelection? {
        guard let data = defaults.data(forKey: key) else { return nil }
        return try? JSONDecoder().decode(ModelSelection.self, from: data)
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/data/AppSettings.kt
package de.maxifrz.lernwerk.data

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import de.maxifrz.lernwerk.llm.ClaudeClient
import de.maxifrz.lernwerk.llm.FailingClient
import de.maxifrz.lernwerk.llm.LlmClient
import de.maxifrz.lernwerk.llm.LlmError
import de.maxifrz.lernwerk.llm.LlmProvider
import de.maxifrz.lernwerk.llm.LlmTask
import de.maxifrz.lernwerk.llm.ModelSelection
import de.maxifrz.lernwerk.llm.OpenAiCompatibleClient
import de.maxifrz.lernwerk.research.WikipediaClient
import de.maxifrz.lernwerk.tutor.DemoLlmClient
import kotlinx.serialization.json.Json

class AppSettings(context: Context) {
    private val prefs = context.getSharedPreferences("settings", Context.MODE_PRIVATE)
    private val keys = ApiKeyStore(context)
    private val transport = OkHttpTransport()

    /** Wikipedia for research in presentations and the study plan; free and without a key. */
    val wikipedia = WikipediaClient(transport)

    /** Model for the help panel and the flashcards created from it. */
    var tutor by mutableStateOf(load(KEY_TUTOR) ?: ModelSelection.default(LlmTask.TUTOR, LlmProvider.NVIDIA))
        private set

    /** Model that reads the material and builds the study plan. */
    var plan by mutableStateOf(load(KEY_PLAN) ?: ModelSelection.default(LlmTask.PLAN, LlmProvider.OPEN_ROUTER))
        private set

    var demoMode by mutableStateOf(prefs.getBoolean(KEY_DEMO, false))
        private set

    var providersWithKey by mutableStateOf(LlmProvider.entries.filter { keys.load(account(it)) != null }.toSet())
        private set

    val hasAnyKey get() = providersWithKey.isNotEmpty()

    fun hasKey(provider: LlmProvider) = provider in providersWithKey

    fun selection(task: LlmTask) = if (task == LlmTask.TUTOR) tutor else plan

    fun setSelection(task: LlmTask, selection: ModelSelection) {
        if (task == LlmTask.TUTOR) tutor = selection else plan = selection
        prefs.edit().putString(if (task == LlmTask.TUTOR) KEY_TUTOR else KEY_PLAN, Json.encodeToString(ModelSelection.serializer(), selection)).apply()
    }

    fun updateDemoMode(enabled: Boolean) {
        demoMode = enabled
        prefs.edit().putBoolean(KEY_DEMO, enabled).apply()
    }

    fun saveKey(key: String, provider: LlmProvider): Boolean {
        val trimmed = key.trim()
        if (trimmed.isEmpty() || !keys.save(account(provider), trimmed)) return false
        providersWithKey = providersWithKey + provider
        // Saving a key means the student wants real answers; a demo mode left on from the sample material would hide them.
        updateDemoMode(false)
        return true
    }

    fun deleteKey(provider: LlmProvider) {
        keys.delete(account(provider))
        providersWithKey = providersWithKey - provider
    }

    fun modelLabel(task: LlmTask): String {
        if (demoMode) return "Demo"
        val chosen = selection(task)
        return chosen.provider.option(chosen.model)?.name ?: chosen.model
    }

    fun makeClient(task: LlmTask): LlmClient {
        if (demoMode) return DemoLlmClient()
        val chosen = selection(task)
        val model = chosen.model.trim()
        if (model.isEmpty()) return FailingClient(LlmError.MissingModel)
        val key = keys.load(account(chosen.provider))
            ?: return FailingClient(LlmError.MissingApiKey(chosen.provider.displayName))
        return when (chosen.provider) {
            LlmProvider.ANTHROPIC -> ClaudeClient(key, model, transport)
            LlmProvider.NVIDIA, LlmProvider.OPEN_ROUTER, LlmProvider.GOOGLE -> OpenAiCompatibleClient(
                provider = chosen.provider,
                apiKey = key,
                model = model,
                sendsImages = chosen.sendsImages,
                transport = transport,
                fallbackModels = listOfNotNull(chosen.provider.fallbackModelIds.firstOrNull { it != model }),
                compressImage = ImageCompressor::jpeg,
            )
        }
    }

    private fun account(provider: LlmProvider) = "llm.${provider.name.lowercase()}"

    private fun load(key: String): ModelSelection? =
        prefs.getString(key, null)?.let { runCatching { Json.decodeFromString(ModelSelection.serializer(), it) }.getOrNull() }

    private companion object {
        const val KEY_TUTOR = "llm.tutor"
        const val KEY_PLAN = "llm.plan"
        const val KEY_DEMO = "demoMode"
    }
}
